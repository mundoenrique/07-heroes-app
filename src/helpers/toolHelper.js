import { NavLink } from 'react-router-dom';
import { heroes } from '../data/heroes';

export const NavBarLink = ({ children, to, className = '', ...props }) => {
	return (
		<>
			<NavLink
				className={({ isActive }) =>
					isActive ? `${className} active` : `${className}`
				}
				to={to}
				{...props}
			>
				{children}
			</NavLink>
		</>
	);
};

export const getHeroesById = (id) => {};

export const getHeroesByPublisher = (publisher) => {
	const validPublishers = ['DC Comics', 'Marvel Comics'];

	if (!validPublishers.includes(publisher)) {
		throw new Error(`${publisher} is not a valid publisher`);
	}

	return heroes.filter((hero) => hero.publisher === publisher);
};
