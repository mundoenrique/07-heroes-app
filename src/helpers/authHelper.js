import { createContext } from 'react';

import { types } from './toolHelper';

export const AuthContext = createContext(null);

export const authReducer = (state = {}, action) => {
	const { login, logout } = types;
	const { type, payload } = action;

	switch (type) {
		case login:
			return {
				...payload,
				logged: true,
			};
		case logout:
			return {
				logged: false,
			};
		default:
			return state;
	}
};
