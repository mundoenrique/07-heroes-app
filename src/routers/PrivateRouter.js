import { useContext } from 'react';
import { Navigate } from 'react-router-dom';

import { AuthContext } from '../helpers/authHelper';

export default function PrivateRouter({ children }) {
	const {
		user: { logged },
	} = useContext(AuthContext);

	return logged ? children : <Navigate to="/login" />;
}
