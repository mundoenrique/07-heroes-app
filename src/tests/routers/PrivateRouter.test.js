import { mount } from 'enzyme';
import { MemoryRouter } from 'react-router-dom';

import { AuthContext } from '../../helpers/authHelper';
import PrivateRouter from '../../routers/PrivateRouter';

jest.mock('react-router-dom', () => ({
	...jest.requireActual('react-router-dom'),
	Navigate: () => <span>Sin acceso</span>,
}));

describe('Pruebas sobre <PrivateRouter.test />', () => {
	Storage.prototype.setItem = jest.fn();

	test('Debe validar la autenticación y escribir en el localstorage', () => {
		const contexValue = {
			user: {
				name: 'Enrique Peñaloza',
				logged: true,
			},
		};

		const wrapper = mount(
			<AuthContext.Provider value={contexValue}>
				<MemoryRouter initialEntries={['/']}>
					<PrivateRouter>
						<h1>Private Route</h1>
					</PrivateRouter>
				</MemoryRouter>
			</AuthContext.Provider>
		);

		expect(wrapper).toMatchSnapshot();
		expect(wrapper.text().trim()).toBe('Private Route');
		expect(localStorage.setItem).toHaveBeenCalledWith('lastPath', '/');
	});

	test('Debe bloquear el componente por falta de autenticación', () => {
		const contexValue = {
			user: {
				logged: false,
			},
		};

		const wrapper = mount(
			<AuthContext.Provider value={contexValue}>
				<MemoryRouter initialEntries={['/']}>
					<PrivateRouter>
						<h1>Private Route</h1>
					</PrivateRouter>
				</MemoryRouter>
			</AuthContext.Provider>
		);

		expect(wrapper.text().trim()).toBe('Sin acceso');
	});
});
