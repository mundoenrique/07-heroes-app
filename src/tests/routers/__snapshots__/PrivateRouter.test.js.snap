// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Pruebas sobre <PrivateRouter.test /> Debe validar la autenticaci√≥n y escribir en el localstorage 1`] = `
<h1>
  Private Route
</h1>
`;
